### PR TITLE
[MH-12820] Mattermost-notification-workflowoperationhandler

### PR DESCRIPTION
**Pull request by isabellakutger**

### CODE REVIEW

File: modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+   * parameter with in between each try a sleep time.
+   *
+   * @param request     The request to execute
+   * @param maxAttempts The number of attempts in case of error
+   * @param timeout     The wait time in milliseconds at which a connection attempt will throw
+   * @return true if the request has been executed successfully
+   */
+  private boolean executeRequest(HttpUriRequest request, int maxAttempts, int timeout, int sleepTime) {
+
+    logger.debug(format("Executing notification request on target %s, %d attemps left", request.getURI(), maxAttempts));
```
Comment (lkiesow) [inline modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java]: No need for `String.format(…)`. The logger supports formatted messages itself:

```Java
logger.debug("Executing notification request on target {}, {} attemps left", request.getURI(), maxAttempts);
```

File: modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java [hunk 2]
```diff
@@ -0,0 +1,295 @@
...
+
+    RequestConfig config = RequestConfig.custom().setConnectTimeout(timeout).setConnectionRequestTimeout(timeout)
+            .setSocketTimeout(timeout).build();
+    CloseableHttpClient httpClient = HttpClientBuilder.create().setDefaultRequestConfig(config).build();
+
+    HttpResponse response;
+    try {
+      response = httpClient.execute(request);
+    } catch (ClientProtocolException e) {
+      logger.error("Protocol error during execution of query on target {}: {}", request.getURI(), e.getMessage());
```
Comment (lkiesow) [inline modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java]: Üöease use the loggers ability to handle exceptions:

```Java
logger.error("Protocol error during execution of query on target {}", request.getURI(), e);
```

File: modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+      return false;
+    } catch (IOException e) {
+      logger.error("I/O error during execution of query on target {}: {}", request.getURI(), e.getMessage());
```
Comment (lkiesow) [inline modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java]: same here

File: modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+      return false;
+    }
+
+    Integer statusCode = response.getStatusLine().getStatusCode();
+    if (statusCode == SC_OK || statusCode == SC_NO_CONTENT || statusCode == SC_ACCEPTED) {
+      logger.debug(format("Request successfully executed on target %s, status code: %d", request.getURI(), statusCode));
```
Comment (lkiesow) [inline modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java]: no need for `format`

File: modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+      return true;
+    } else if (maxAttempts > 1) {
+      logger.debug(format("Request failed on target %s, status code: %d, will retry in %d seconds", request.getURI(),
+              statusCode, sleepTime / 1000));
```
Comment (lkiesow) [inline modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java]: same here (I'm just ignoring this in my review from here on)

File: modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+      try {
+        Thread.sleep(sleepTime);
+        return executeRequest(request, --maxAttempts, timeout, sleepTime * SLEEP_SCALE_FACTOR);
+      } catch (InterruptedException e) {
+        logger.error("Error during sleep time before new notification request try: {}", e.getMessage());
+        return false;
+      }
+    } else {
+      logger.warn("Request failed on target {}, status code: {}, no more attempt.", request.getURI(), statusCode);
+      return false;
```
Comment (lkiesow) [inline modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java]: Didn't follow the whole flow but this looks like it should cause an error

File: modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+   * Checks if an object is null. If an object is null, then method returns not defined, else it returns object as a
+   * String
+   *
+   * @param o The object to check
+   * @param s The name of metadata to check
+   * @return
+   */
+  private String checkIfNull(Object o, String s) {
+
+    if (o != null) {
```
Comment (lkiesow) [inline modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java]: Reversing this might make it easier to read:

```Java
if (o != null) {
  return s + " not defined ";
}
if (o instanceof String[]) {
  …
}
return o.toString();
```

File: modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+    s = s.replace("%I", checkIfNull(workflowInstance.getMediaPackage().getIdentifier(), "Mediapackage-ID"));
+    s = s.replace("%T", checkIfNull(workflowInstance.getMediaPackage().getTitle(), "Mediapackage-Title"));
+    s = s.replace("%c", checkIfNull(workflowInstance.getMediaPackage().getContributors(), "Contributurs"));
+    s = s.replace("%C", checkIfNull(workflowInstance.getMediaPackage().getCreators(), "Creators"));
+    s = s.replace("%D", checkIfNull(workflowInstance.getMediaPackage().getDate(), "Date"));
+    s = s.replace("%d", checkIfNull(workflowInstance.getMediaPackage().getDuration(), "Duration"));
+    s = s.replace("%l", checkIfNull(workflowInstance.getMediaPackage().getLanguage(), "Language"));
+    s = s.replace("%L", checkIfNull(workflowInstance.getMediaPackage().getLicense(), "License"));
+    s = s.replace("%S", checkIfNull(workflowInstance.getMediaPackage().getSeriesTitle(), "Series-Title"));
+    String json = "{\"text\": \"" + s + "\"}";
```
Comment (lkiesow) [inline modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java]: This would break r.g. if one of the metadata fields contain quotes. Please use a JSON library to handle this properly. I recommend [GSON](http://tutorials.jenkov.com/java-json/gson.html#generating-json-from-java-objects) which is amazingly easy to use on collections and is used in Opencast anyway.

File: modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java [hunk 9]
```diff
@@ -0,0 +1,295 @@
...
+
+  /**
+   * Name of the subject HTTP parameter
+   */
+  public static final String HTTP_PARAM_PAYLOAD = "payload";
+
+  /**
+   * The configuration options for this handler
+   */
+  private static final SortedMap<String, String> CONFIG_OPTIONS;
```
Comment (lkiesow) [inline modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java]: This (and the block below) should be unnecessary by now

File: docs/guides/admin/docs/modules/notification.md [hunk 10]
```diff
@@ -0,0 +1,37 @@
+# Mattermost Notification
+#### Example for mattermost-notify operation:
```
Comment (lkiesow) [inline docs/guides/admin/docs/modules/notification.md]: You shouldn't skip a header level. It should be something like:

```
Header Level One
==============

Header Level Two
------------------------
```

…or it's equivalent using `#` and `##`

File: docs/guides/admin/docs/modules/notification.md [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+~~~~
```
Comment (lkiesow) [inline docs/guides/admin/docs/modules/notification.md]:
```XML
    …
    ```

File: docs/guides/admin/docs/modules/notification.md [hunk 12]
```diff
@@ -0,0 +1,37 @@
+# Mattermost Notification
```
Comment (lkiesow) [inline docs/guides/admin/docs/modules/notification.md]: A short description is missing. What is this operation for?

Comment (lkiesow) [reply docs/guides/admin/docs/modules/notification.md]: …and the list of configuration options is missing

File: modules/mattermost-notification-workflowoperation/pom.xml [hunk 13]
```diff
@@ -0,0 +1,99 @@
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+  <modelVersion>4.0.0</modelVersion>
+  <artifactId>mattermost-notification-workflowoperation</artifactId>
+  <packaging>bundle</packaging>
+  <name>mattermost-notification-workflowoperation</name>
+  <parent>
+    <groupId>org.opencastproject</groupId>
+    <artifactId>mattermost-notification</artifactId>
+    <version>3.0-SNAPSHOT</version>
```
Comment (lkiesow) [inline modules/mattermost-notification-workflowoperation/pom.xml]: This should be `${project.version}`

File: modules/mattermost-notification-workflowoperation/pom.xml [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <version>${junit.version}</version>
+      <scope>test</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.easymock</groupId>
+      <artifactId>easymock</artifactId>
+      <version>3.4</version>
```
Comment (lkiesow) [inline modules/mattermost-notification-workflowoperation/pom.xml]: Versions are specified in main pom.xml

File: modules/mattermost-notification-workflowoperation/pom.xml [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+      <version>${project.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>org.opencastproject</groupId>
+      <artifactId>matterhorn-search-service-api</artifactId>
+      <version>${project.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>org.opencastproject</groupId>
+      <artifactId>matterhorn-email-template-service-api</artifactId>
```
Comment (lkiesow) [inline modules/mattermost-notification-workflowoperation/pom.xml]: What do you need this for?

**Review by lkiesow — CHANGES_REQUESTED** [review]
Most of the things I commented on should be requests for minor changes. There does not seem to be an obvious big issue (never tested it though ;-P)

Apart from the comments:

- This should be in the main pom.xml to be build
- The documentation should be linked from the list of workflow operation handler and needs to be added to `mkdocs.yml`
- If possible, there should be some unit tests for this code

File: modules/mattermost-notification-workflowoperation/src/main/resources/OSGI-INF/operations/mattermost-notify.xml [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+               name="org.opencastproject.workflow.handler.mattermost.notification.MattermostNotificationWorkflowOperationHandler"
+               immediate="true">
+  <implementation
+      class="org.opencastproject.workflow.handler.mattermost.notification.MattermostNotificationWorkflowOperationHandler"/>
+  <property name="service.description" value="Mattermost Notification Operation Handler"/>
+  <property name="workflow.operation" value="mattermost-notify"/>
+  <service>
+    <provide interface="org.opencastproject.workflow.api.WorkflowOperationHandler"/>
+  </service>
+  <reference name="ServiceRegistry" cardinality="1..1"
```
Comment (gregorydlogan) [inline modules/mattermost-notification-workflowoperation/src/main/resources/OSGI-INF/operations/mattermost-notify.xml]: This isn't used as far as I can tell.

**Comment by gregorydlogan** [comment]
Is there a JIra ticket associated with this?

File: modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java (hunk 6 above)
Comment (isabellakutger) [reply modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java]: Yes it causes an exception which is thrown in line 188

**Comment by staubesv** [comment]
@isabellakutger Isn't this WOH almost the same as WOH http-notify? (see https://docs.opencast.org/develop/admin/workflowoperationhandlers/httpnotify-woh/)

Please also note that "Mattermost" is not a suitable name as the community spent quite some time to get rid of the name "Matterhorn".

**Comment by JamesUoM** [comment]
@staubesv "mattermost" is the name of a Slack like notification appl https://about.mattermost.com/ it is just coincidence it sounds a bit like our old moniker.

**Comment by staubesv** [comment]
@isabellakutger Oh, after I've learned that Mattermost is a service, I've now got the point why this is different from WOH http-notify, too ;-) @JamesUoM: Thanks for the info!

**Comment by isabellakutger** [comment]
@gregorydlogan no, there is no JIra ticket associated with this.

**Comment by staubesv** [comment]
@isabellakutger Note that pull requests are supposed to be accompanied by JIRA tickets (see https://docs.opencast.org/develop/developer/development-process/). That's why @gregorydlogan was asking... Please create JIRA tickets (with reasonable descriptions) for your PRs.

File: modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java [hunk 17]
```diff
@@ -0,0 +1,284 @@
...
+   * Checks if an object is null. If an object is null, then method returns not defined, else it returns object as a
+   * String
+   *
+   * @param o The object to check
+   * @param s The name of metadata to check
+   * @return
+   */
+  private String checkIfNull(Object o, String s) {
+
+    if(o != null) {
```
Comment (mtneug) [inline modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java]: You did not change the logic to `o == null`.

File: modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java [hunk 18]
```diff
@@ -0,0 +1,284 @@
...
+    s = s.replace("%C", checkIfNull(workflowInstance.getMediaPackage().getCreators(), "Creators"));
+    s = s.replace("%D", checkIfNull(workflowInstance.getMediaPackage().getDate(), "Date"));
+    s = s.replace("%d", checkIfNull(workflowInstance.getMediaPackage().getDuration(), "Duration"));
+    s = s.replace("%l", checkIfNull(workflowInstance.getMediaPackage().getLanguage(), "Language"));
+    s = s.replace("%L", checkIfNull(workflowInstance.getMediaPackage().getLicense(), "License"));
+    s = s.replace("%S", checkIfNull(workflowInstance.getMediaPackage().getSeriesTitle(), "Series-Title"));
+    GsonBuilder builder = new GsonBuilder();
+    Gson gson = builder.create();
+    String json = gson.toJson(s);
+    return json;
```
Comment (mtneug) [inline modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java]: This does not produce a valid JSON string (at least for Slack). It just adds quotes around the string and adds escape symbols. To create the required JSON object, you have to execute something like this:

```java
JsonObject json = new JsonObject();
json.addProperty("text", s);

GsonBuilder builder = new GsonBuilder();
Gson gson = builder.create();
return gson.toJson(json);
```

File: docs/guides/admin/docs/modules/notification.md (hunk 12 above)
Comment (staubesv) [reply docs/guides/admin/docs/modules/notification.md]: @isabellakutger Just wondering: Why are you actually adding a new WOH for notification at all (instead of just enhancing the existing HTTPNotifyWorkflowOperationHandler)?  At the end of the day, the description suggests that this WOH is not specifically built for Mattermost and in overall, it looks like HTTPNotify with some more metadata templating.

In case you are looking for fancy ideas: The EmailNotficationWorkflowHandler comes with a way more powerful metadata templating mechanism. Maybe it could be re-used?

Comment (isabellakutger) [reply docs/guides/admin/docs/modules/notification.md]: @staubesv it was something I was told to implement to get to know how WOH works. So maybe there could be a more suitable way to do this.

If I have time, I'll look at the EmailNotificationWorkflowHandler and try to use it instead.

**Comment by staubesv** [comment]
Hi @isabellakutger I've "officially" took the review ;-)

Will have a closer look next week and report back here.

File: docs/guides/admin/docs/modules/notification.md [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+## Description
+The MattermostNotificationOperationHander sends an notification to a channel of Mattermost or similar applications,
```
Comment (staubesv) [inline docs/guides/admin/docs/modules/notification.md]: typo: "sends **a** notification"

File: docs/guides/admin/docs/modules/notification.md [hunk 20]
```diff
@@ -0,0 +1,55 @@
...
+like Slack, with the chosen parameters provided. It is useful to send such notifications when some operation(s) have
+been completed or some error has occurred in a workflow.
+The notification message can be freely chosen. You can use different parameters which will be replaced with the
+corresponding meta-data of the current workflow instance (see List of parameters).
```
Comment (staubesv) [inline docs/guides/admin/docs/modules/notification.md]: typo: metadata

File: docs/guides/admin/docs/modules/notification.md [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+
+## List of configuration options
+
+configuration keys | description |default
```
Comment (staubesv) [inline docs/guides/admin/docs/modules/notification.md]: This is not correct markdown.

Reproduction:
cd docs/guides/admin
mkdocs serve

Have a look at this table ;-)

Please also try to make this table more readable in plain text ("formatting" using spaces)

File: docs/guides/admin/docs/modules/notification.md [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+
+
+## Example for mattermost-notify operation
+
+```XML
+  <operation
+      id="mattermost-notify"
+      fail-on-error="false"
+      exception-handler-workflow="error"
+      description="Notify mattermost about error">
```
Comment (staubesv) [inline docs/guides/admin/docs/modules/notification.md]: Typo: **M**attermost

File: docs/guides/admin/docs/modules/notification.md [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+        <configuration key="message">Error at Workflow %i (%t) State: %s</configuration>
+        <configuration key="method">post</configuration>
+        <configuration key="max-retry">3</configuration>
+        <configuration key="timeout">5</configuration>
+      </configurations>
+    </operation>
+```
+
+## List of parameters
+All parameters (%<letter>) will be substituted with corresponding meta-data of current workflow instance.
```
Comment (staubesv) [inline docs/guides/admin/docs/modules/notification.md]: ... **metadata** of **the** current workflow instance

File: docs/guides/admin/docs/modules/notification.md [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+
+Parameter | Meta-Data
```
Comment (staubesv) [inline docs/guides/admin/docs/modules/notification.md]: **Metadata**

File: docs/guides/admin/mkdocs.yml [hunk 25]
```diff
@@ -63,6 +63,7 @@ pages:
    - Execute Service: 'modules/execute.md'
    - Live Schedule: 'modules/liveschedule.md'
    - LTI Module: 'modules/ltimodule.md'
+   - Mattermost Notification Module: 'modules/notification.md'
```
Comment (staubesv) [inline docs/guides/admin/mkdocs.yml]: That is not a module but rather a workflow operation handler and there belong to the section "Workflow Operation Handlers" (and the file into the folder "workflowoperationhandler")

File: modules/mattermost-notification-workflowoperation/pom.xml [hunk 26]
```diff
@@ -0,0 +1,95 @@
...
+      <groupId>org.opencastproject</groupId>
+      <artifactId>matterhorn-workflow-service-api</artifactId>
+      <version>${project.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>org.opencastproject</groupId>
+      <artifactId>matterhorn-search-service-api</artifactId>
+      <version>${project.version}</version>
+    </dependency>
+    <dependency>
```
Comment (staubesv) [inline modules/mattermost-notification-workflowoperation/pom.xml]: While unit tests would be desirable, this PR currently does not provide them. You should at least remove the depenencies junit and easymock here.

File: modules/mattermost-notification-workflowoperation/pom.xml [hunk 27]
```diff
@@ -0,0 +1,95 @@
...
+      <groupId>org.opencastproject</groupId>
+      <artifactId>matterhorn-kernel</artifactId>
+      <version>${project.version}</version>
+    </dependency>
+    <dependency>
+      <groupId>org.opencastproject</groupId>
+      <artifactId>matterhorn-workflow-service-api</artifactId>
+      <version>${project.version}</version>
+    </dependency>
+    <dependency>
```
Comment (staubesv) [inline modules/mattermost-notification-workflowoperation/pom.xml]: Does this depend on the search service?

File: modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java [hunk 28]
```diff
@@ -0,0 +1,286 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Workflow operation for notifying mattermost about status of current workflow.
```
Comment (staubesv) [inline modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java]: **M**attermost (it is a product name, isn't it?)

... about **the** status of **the** current ...

File: modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java [hunk 29]
```diff
@@ -0,0 +1,286 @@
...
+   */
+  private static final Logger logger = LoggerFactory.getLogger(MattermostNotificationWorkflowOperationHandler.class);
+
+  /**
+   * Default value for the number of attempts for a request
+   */
+  private static final int DEFAULT_MAX_RETRY = 5;
+
+  /**
+   * Default maximum wait time the client when trying to execute a request
```
Comment (staubesv) [inline modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java]: Please specify unit (milliseconds?)

File: modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java [hunk 30]
```diff
@@ -0,0 +1,286 @@
...
+   */
+  private static final int DEFAULT_TIMEOUT = 10 * 1000;
+
+  /**
+   * Default time between two request attempts
```
Comment (staubesv) [inline modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java]: Please specify unit (milliseconds?)

File: modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java [hunk 31]
```diff
@@ -0,0 +1,286 @@
...
+    int timeout = DEFAULT_TIMEOUT;
+
+    // Required configuration
+    String urlPath = getConfig(workflowInstance, OPT_URL_PATH);
+
+    // Optional configuration
+    String notificationMessage = getConfig(workflowInstance, OPT_NOTIFICATION_MESSAGE, null);
+    String method = getConfig(workflowInstance, OPT_METHOD, "post");
+    String maxRetryOpt = getConfig(workflowInstance, OPT_MAX_RETRY, null);
+    String timeoutOpt = getConfig(workflowInstance, OPT_TIMEOUT, null);
```
Comment (staubesv) [inline modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java]: You don't seem to set the default for maxRetryOpt and timeoutOpt although you have defined constants for those...

File: modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java [hunk 32]
```diff
@@ -0,0 +1,286 @@
...
+    }
+
+    // Is there a need to retry on failure?
+    if (maxRetryOpt != null) {
+      maxRetry = Integer.parseInt(maxRetryOpt);
+    }
+
+    // Figure out which request method to use
+    HttpEntityEnclosingRequestBase request = null;
+    if (StringUtils.equalsIgnoreCase("post", method)) {
```
Comment (staubesv) [inline modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java]: Please add string constants for "post" and "put"

**Review by staubesv — CHANGES_REQUESTED** [review]

**Comment by staubesv** [comment]
Hi @isabellakutger 

Sorry for the delay... I'm just back from three weeks vacation. The PR looks good to me but now has a merge conflict.

Please fix the merge conflict... then it is ready to go.

**Comment by staubesv** [comment]
@isabellakutger 

Just wanted to merge this but found some last issues...

The new module mattermost-notify this PR adds is currently neither part of the main pom nor part of any feature.

You need to add this module to the main pom (https://github.com/opencast/opencast/blob/develop/pom.xml) so that it is actually compiled when building Opencast.

You also need to add this module to the appropriate features (https://github.com/opencast/opencast/blob/develop/assemblies/karaf-features/src/main/feature/feature.xml). You can add it to the same features (all-in-one, admin, adminpresenter and adminworker) as the module "opencast-notification-workflowoperation" that contains very similar workflow operations.

Alternatively, you could also put this WOH into that module instead of introducing a new one (no opinion from my side, though).

File: modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java [hunk 33]
```diff
@@ -0,0 +1,297 @@
...
+
+  }
+
+  /**
+   * Execute the given notification request. If the target is not responding, retry as many time as the maxAttampts
+   * parameter with in between each try a sleep time.
+   *
+   * @param request     The request to execute
+   * @param maxAttempts The number of attempts in case of error
+   * @param timeout     The wait time in milliseconds at which a connection attempt will throw
```
Comment (staubesv) [inline modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java:257]: sleepTime is missing

**Comment by staubesv** [comment]
BTW: I couldn't reproduce mvn site failing locally, but could you anyway add the missing "sleepTime" argument (see separate comment) and add the missing descriptions for all @return?

**Comment by staubesv** [comment]
@isabellakutger I assume that you have realised that you messed something with your latest commit and are working on this. Please let me know once I shall continue the review.

File: modules/mattermost-notification-workflowoperation/${project.build.directory}/classes/OSGI-INF/operations/mattermost-notify.xml [hunk 34]
```diff
@@ -0,0 +1,15 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (staubesv) [inline modules/mattermost-notification-workflowoperation/${project.build.directory}/classes/OSGI-INF/operations/mattermost-notify.xml]: This file does also not belong to this PR

**Comment by staubesv** [comment]
@isabellakutger Please note that the feature freeze for Opencast 6.0 is on September 25th and this PR still has issues that need to be resolved.

... so if you want to get this into 6.0, you should resolve the issues "soon"

**Comment by staubesv** [comment]
@isabellakutger 

Please note that the PR is still missing two crucial things:

1. You need to include the new module mattermost-notification-workflowoperation in the main POM. Without this, the module does not even get compiled when building Opencast. I'm talking about [this](https://github.com/opencast/opencast/blob/develop/pom.xml) file.
2. You need to include the module in [here](https://github.com/opencast/opencast/blob/develop/assemblies/karaf-features/src/main/feature/feature.xml). If you don't add the module there, the workflow operation handler won't be available in a deployed system. You can just add it at the same locations where the module notification-workflowoperation appears

Please let me know in case it is not clear what needs to be done here.

File: modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java [hunk 35]
```diff
@@ -0,0 +1,298 @@
...
+   * Checks if an object is null. If an object is null, then method returns not defined, else it returns object as a
+   * String
+   *
+   * @param o The object to check
+   * @param s The name of metadata to check
+   * @return String containing the transformed object
+   */
+  private String checkIfNull(Object o, String s) {
+
+    if(o == null) {
```
Comment (staubesv) [inline modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java]: That results in a build error:

[ERROR] /home/travis/build/opencast/opencast/modules/mattermost-notification-workflowoperation/src/main/java/org/opencastproject/workflow/handler/mattermost/notification/MattermostNotificationWorkflowOperationHandler.java:239:7: 'if' is not followed by whitespace. [WhitespaceAround]
Audit done.

**Review by staubesv — APPROVED** [review]

**Comment by staubesv** [comment]
@isabellakutger Thanks for this nice feature!
